crates/mun_compiler/src/diagnostics.rs

```diff
@@ -3,111 +3,510 @@ use mun_hir::{FileId, HirDatabase, HirDisplay, Module};
 use mun_syntax::{ast, AstNode, SyntaxKind};
 use std::cell::RefCell;
 
-mod emit;
+use annotate_snippets::snippet::{Annotation, AnnotationType, Slice, Snippet, SourceAnnotation};
 
-pub use emit::Emit;
-use mun_errors::{Diagnostic, Level};
+use unicode_segmentation::UnicodeSegmentation;
+
+/// Replace all tabs('\t') in string to spaces(' ')
+fn replace_tabs(text: &mut String) {
+    let mut tab_index = text.find('\t');
+    while let Some(index) = tab_index {
+        #[allow(clippy::range_plus_one)]
+        text.replace_range(index..index + 1, " ");
+        tab_index = text.find('\t');
+    }
+}
+
+/// Returns amount of graphemes in text range
+fn graphemes_amount_in_range(range: (usize, usize), text: &str) -> usize {
```

Comment:
Maybe rename to `count_graphemes_in_range`

Reply:
Why not use `text[range.0, range.1]` ?

This entire function can be replaced with `UnicodeSegmentation::graphemes(text[range], true).count()` where range is the Range that you want to check.

See: https://doc.rust-lang.org/1.30.0/std/primitive.str.html#impl-Index%253CRange%253Cusize%253E%253E